src/handle.rs

```diff
@@ -432,3 +449,301 @@ impl HandleHost {
+  data: Option<Box<WeakData<T>>>,
+  isolate_handle: IsolateHandle,
+}
+
+impl<T> Weak<T> {
+  pub fn new(isolate: &mut Isolate, handle: impl Handle<Data = T>) -> Self {
+    let HandleInfo { data, host } = handle.get_handle_info();
+    host.assert_match_isolate(isolate);
+    Self::new_raw(isolate, data, None)
+  }
+
+  /// Create a weak handle with a finalization callback installed.
+  ///
+  /// NOTE: There is no guarantee as to *when* or even *if* the callback is
+  /// invoked. The invocation is performed solely on a best effort basis. As
+  /// always, GC-based finalization should *not* be relied upon for any critical
+  /// form of resource management!
+  ///
+  /// The callback does not have access to the inner value, because it has
+  /// already been collected by the time it runs.
+  pub fn with_finalizer(
+    isolate: &mut Isolate,
+    handle: impl Handle<Data = T>,
+    finalizer: Box<dyn FnOnce()>,
+  ) -> Self {
+    let HandleInfo { data, host } = handle.get_handle_info();
+    host.assert_match_isolate(isolate);
+    Self::new_raw(isolate, data, Some(finalizer))
+  }
+
+  fn new_raw(
+    isolate: *mut Isolate,
+    data: NonNull<T>,
+    finalizer: Option<Box<dyn FnOnce()>>,
+  ) -> Self {
+    let weak_data = Box::new(WeakData {
+      pointer: Default::default(),
+      finalizer: Cell::new(finalizer),
+      weak_dropped: Cell::new(false),
+    });
+    let data = data.cast().as_ptr();
+    let data = unsafe {
+      v8__Global__NewWeak(
+        isolate,
+        data,
+        weak_data.deref() as *const _ as *const c_void,
+        Self::first_pass_callback,
+      )
+    };
+    weak_data
+      .pointer
+      .set(Some(unsafe { NonNull::new_unchecked(data as *mut _) }));
+    Self {
+      data: Some(weak_data),
+      isolate_handle: unsafe { (*isolate).thread_safe_handle() },
+    }
+  }
+
+  /// Creates a new empty handle, identical to one for an object that has
+  /// already been GC'd.
+  pub fn empty(isolate: &mut Isolate) -> Self {
+    Weak {
+      data: None,
+      isolate_handle: isolate.thread_safe_handle(),
+    }
+  }
+
+  /// Clones this handle and installs a finalizer callback on the clone, as if
+  /// by calling [`Self::with_finalizer`].
+  ///
+  /// Note that if this handle is empty (its value has already been GC'd), the
+  /// finalization callback will never run.
+  pub fn clone_with_finalizer(&self, finalizer: Box<dyn FnOnce()>) -> Self {
+    self.clone_raw(Some(finalizer))
+  }
+
+  fn clone_raw(&self, finalizer: Option<Box<dyn FnOnce()>>) -> Self {
+    if let Some(data) = self.get_pointer() {
+      Self::new_raw(
+        // SAFETY: We're in the isolate's thread, because Weak<T> isn't Send or
+        // Sync.
+        unsafe { self.isolate_handle.get_isolate_ptr() },
+        data,
+        finalizer,
+      )
+    } else {
+      Weak {
+        data: None,
+        isolate_handle: self.isolate_handle.clone(),
+      }
+    }
+  }
+
+  fn get_pointer(&self) -> Option<NonNull<T>> {
```

Comment:
In NAPI, we implement references to these Weak, if the ref count is greater than 0 we need to "clear the weak" and prevent finalization (sorta like what's happening in `drop()` rn?) What would be the right API for this?

Reply:
Per conversation on Discord, keeping track of the reference count and clearing the weak handle can be done in Deno, but we would need a way to prevent finalization of a `Weak`, possibly even after the first pass callback has run. Since the second pass callback will only run the finalizer if it's `Some`, preventing finalization would just involve setting it to `None`.